File: lab02/CoinFlipper.java
import java.util.Scanner;

/******************************************************************************************************************
 * CoinFlipper.java
 * Adapted from Dean & Dean - Introduction to Programming with Java: A Problem Solving Approach, ch 9.
 *
 * This program generates a histogram of coin flips. Suppose you have 5 coins and flip them 100000 times. 
 * Some flips will result in all heads, some with all tails, and others with some combination of heads and tails.
 * The histogram displays the number of times each combination occurred over all flips.
 ******************************************************************************************************************/
public class CoinFlipper {

    /**
     * Program execution begins here
     * @param args not used
     */
    public static void main(String[] args) {
        CoinFlipper coinFlipper = new CoinFlipper();
        coinFlipper.run();
    }

    /**
     * CoinFlipper - Constructor
     */
    private CoinFlipper() {
        // Intentionally left blank
    }

    /**
     * Gathers input from the user,
     * controls execution, and measures how long it takes to execute.
     */
    private void run() {
        int numberOfCoins = 0;        // number of coins to flip
        int numberOfRepetitions = 0;  // number of times the coins are flipped

        // The frequency array holds the number of times a particular number of heads occurred:
        // frequency[0] holds the number of times 0 heads occurred
        // frequency[1] holds the number of times 1 head occurred
        // ...
        // frequency[numberOfCoins] holds the number of times all heads occurred
        int[] frequency;

        Scanner kbd = new Scanner(System.in);
        boolean validInput = false;
        while( !validInput ) {
            try {
                System.out.print("Enter the number of coins to be flipped: ");
                numberOfCoins = kbd.nextInt();
                System.out.print("Enter the number of flips: ");
                numberOfRepetitions = kbd.nextInt();
            } catch( Exception e ) {
                System.out.println("Invalid input. Try again.");
                kbd.nextLine(); // consume remaining input
                continue;
            }
            validInput = true;
        }

        long executionTime = System.currentTimeMillis(); // current system time snapshot
        frequency = flipCoins(numberOfCoins, numberOfRepetitions); // flip numberOfCoins coins numberOfRepetitions times
        executionTime = System.currentTimeMillis() - executionTime; // compute elapsed time

        // NOTE: Do not include histogram generation in execution time calculation - console I/O takes a long time compared to internal math computations
        printHistogram(numberOfCoins, numberOfRepetitions, frequency); // display the resulting histogram

        System.out.println("Coin Flipper Time: " + executionTime + "ms");
    }

    /**
     * This method flips a specified number of coins a specified number of times
     * and gathers the number of times a certain number of heads occurred in each flip into a frequency[] array.
     * @param coins the number of coins to flip
     * @param times the number of times to flip each coin
     * @return array representing the frequency of 'heads' for each flip repetition
     */
    private int[] flipCoins(int coins, int times) {
        // This loop fills up frequency 'bins'.
        //    Each iteration simulates one group of coin flips.
        // For example, with a group of flips of 3 coins, heads may come up 0, 1, 2, or 3 times.

        // Allocate the array per user-specified input (note: Java initializes the contents to 0).
        int[] frequency = new int[coins + 1];
        for(int rep = 0; rep < times; rep++) {
            // perform a flip of the coins
            int heads = flipCoinsOneTime(coins);
            frequency[heads]++; // update appropriate bin
        }
        return frequency;
    }

    /**
     * This method flips a set of coins and returns the number heads that occurred.
     * It makes use of a random number generator to randomly generate heads or tails for each flip.
     * @param coins the number of coins to flip
     * @return the number of heads that occurred in the flips
     */
    private int flipCoinsOneTime(int coins) {
        int heads = 0;
        for(int i = 0; i < coins; i++) { // flip the coin
            heads += (int)(Math.random() * 2); // computed random int value is either 0 or 1 (tails or heads)
        }
        return heads; // number of heads that came up
    }

    /**
     * This method prints a histogram of the number of heads that occurred for a specified number of flip repetitions
     * Notes: The output generated for coins=5 and times=100000 may look something like this:
     *
     * Number of times each head count occurred in 100000 flips of 5 coins:
     * 0  3076  ***
     * 1  15792  ****************
     * 2  31348  *******************************
     * 3  31197  *******************************
     * 4  15552  ****************
     * 5  3035  ***
     *
     * @param coins the number of coins flipped in each repetition
     * @param times the number of times the coins were flipped
     * @param frequency the frequency (count) of heads for each flip repetition
     */
    private void printHistogram(int coins, int times, int[] frequency) {
        System.out.println("Number of times each head count occurred in " + times + " flips of " + coins + " coins:");

        // This loop prints the histogram. Each iteration prints one histogram bar.
        for(int heads = 0; heads < frequency.length; heads++) {

            // Compute the length of the histogram bar
            //   Normalize based on the number of repetitions
            double fractionOfReps = (float) frequency[heads] / times;
            int numOfAsterisks = (int) Math.round(fractionOfReps * 100);

            // Print the histogram bar
            System.out.print( " " + heads + "  " + frequency[heads] + "  " );
            for(int i = 0; i < numOfAsterisks; i++) {
                System.out.print("*");
            }
            System.out.println();
        }
    }
}

File: lab02/CoinFlipper.js
/**
 * Lab02 : JavaScript Coin Flipper
 * Name: Paige Rosynek
 * Section: SE 2840 031
 */


/**
 * Runs the coin flipper simulation.
 * Prompts the user for input, runs simulation, and measures how long
 * the simulation takes to execute.
 */
const run = () => {
    // note: prompt returns string
    let numberOfCoins = prompt('Enter the number of coins to be flipped : ', '1');
    let numberOfFlips = prompt('Enter the number of flips : ', '100');

    const isValidInput = checkValidInput(numberOfCoins, numberOfFlips)

    if (isValidInput.value) {
        numberOfCoins = parseInt(numberOfCoins);
        numberOfFlips = parseInt(numberOfFlips);
    }
    else {
        alert(isValidInput.message);
        return;  // exit function
    }

    const startTime = performance.now();
    const frequency = flipCoins(numberOfCoins, numberOfFlips);
    const executionTime = performance.now() - startTime;

    printHistogram(numberOfCoins, numberOfFlips, frequency);

    console.log('Coin flipper time : %f ms', executionTime);
}


/**
 * Simulates the flipping of a given number of coins a specified number of times.
 * The frequency of the number of times a certain number of heads occured is stored in an array.
 * @param coins the number of coins to flip
 * @param flips the number of times to flip each coin
 * @returns {any[]} array representing the frequency of 'heads' for each flip repetition
 */
function flipCoins(coins, flips) {
    let frequency = Array(coins + 1).fill(0);           // initialize array of zeros

    for (let flip = 0; flip < flips; flip++) {
        let heads = flipCoinsOneTime(coins);
        frequency[heads]++;                                             // update appropriate bin
    }

    return frequency;
}


/**
 * Simulates the flipping of a given number of coins one time.
 * The number of heads that occured is returned.
 * @param coins the number of coins to flip
 * @returns {number} the number of heads that occured
 */
function flipCoinsOneTime(coins) {
    let heads = 0;

    for (let i = 0; i < coins; i++) {
        heads += Math.floor(Math.random() * 2);
    }

    return heads;
}


/**
 * Checks if the number of coins and flips is a valid integer and
 * returns whether the values are valid. Returns a literal object that has properties value and
 * message to represent whether the input was valid and the associated error message if value is false.
 *
 * Coins must be in the range [1-10]
 * Flips must be in the range [1-1000000]
 * @param coins the number of coins to flip
 * @param flips the number of times to flip each coin
 * @returns {{message: string, value: boolean}} literal object representing whether the given values were valid and an associated error message
 */
function checkValidInput(coins, flips) {
    const isValid = {
        value : false,
        message : ''                        // error message
    };

    coins = Number(coins);
    flips = Number(flips);

    if (isNaN(coins) || isNaN(flips)) {
        isValid.message = 'Invalid input. Value must be a number.';
    }
    else if (isFloat(coins) || isFloat(flips)) {
        isValid.message = 'Invalid input. Value must be an integer.';
    }
    else if ((coins <= 0) || (coins > 10)) {
        isValid.message = 'Invalid input. Coins must be in the range 1 - 10.';
    }
    else if ((flips <= 0) || (flips > 1000000)) {
        isValid.message = 'Invalid input. Flips must be in the range 1 - 1000000.';
    }
    else {
        isValid.value = true;
    }

    return isValid;
}


/**
 * Determines whether a number is a float.
 * Returns true if n is a float, false otherwise.
 * @param n number to check the type of
 * @returns {boolean} whether or not n is a float (true if float, false otherwise)
 */
function isFloat(n) {
    return !isNaN(n) && n % 1 !== 0;
}


/**
 * Prints a histogram of the number of heads that occurred for the specified number of flip repetitions
 * @param coins the number of coins to flip
 * @param flips the number of times to flip each coin
 * @param frequency array representing the frequency of 'heads' for each flip repetition
 */
function printHistogram(coins, flips, frequency) {
    console.log(`Number of times each head count occurred in ${flips} flips of ${coins} coins : `)

    // print the histogram
    // (freq, index) = (frequency of count, number of heads)
    frequency.forEach((freq, index) => {
        const fraction = (freq / flips);
        const numAstrisks = Math.round(fraction * 100);
        printHistBar(index, numAstrisks, freq);
    });
}


/**
 * Prints a single row of the histogram with the number of heads that occurred,
 * the frequency of the count, and the distribution bar.
 * @param numHeads the number of heads; heads count
 * @param nAstrisks the calculated number of astricks to be printed
 * @param freq the frequency of the head count from out of all flips
 */
function printHistBar(numHeads, nAstrisks, freq) {
    const astricks = Array(nAstrisks).fill('*').join('');
    const bar = ` ${numHeads}  ${freq}  ${astricks}`
    console.log(bar)
}


// Run the coin flipper code when the browser finishes loading the js file
run();
